runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/getdata/DirectGetDataClient.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.dataflow.worker.windmill.client.getdata;
+
+import java.util.function.Function;
+import org.apache.beam.runners.dataflow.worker.WorkItemCancelledException;
+import org.apache.beam.runners.dataflow.worker.windmill.Windmill;
+import org.apache.beam.runners.dataflow.worker.windmill.client.WindmillStream.GetDataStream;
+import org.apache.beam.sdk.annotations.Internal;
+
+/** {@link GetDataClient} that fetches data directly from a specific {@link GetDataStream}. */
+@Internal
+public final class DirectGetDataClient implements GetDataClient {
+
+  private final GetDataStream directGetDataStream;
+  private final Function<String, GetDataStream> sideInputGetDataStreamFactory;
+  private final ThrottlingGetDataMetricTracker getDataMetricTracker;
+
+  private DirectGetDataClient(
+      GetDataStream directGetDataStream,
+      Function<String, GetDataStream> sideInputGetDataStreamFactory,
+      ThrottlingGetDataMetricTracker getDataMetricTracker) {
+    this.directGetDataStream = directGetDataStream;
+    this.sideInputGetDataStreamFactory = sideInputGetDataStreamFactory;
+    this.getDataMetricTracker = getDataMetricTracker;
+  }
+
+  public static GetDataClient create(
+      GetDataStream getDataStream,
+      Function<String, GetDataStream> sideInputGetDataStreamFactory,
+      ThrottlingGetDataMetricTracker getDataMetricTracker) {
+    return new DirectGetDataClient(
+        getDataStream, sideInputGetDataStreamFactory, getDataMetricTracker);
+  }
+
+  @Override
+  public Windmill.KeyedGetDataResponse getStateData(
+      String computation, Windmill.KeyedGetDataRequest request) {
+    if (directGetDataStream.isShutdown()) {
```

Comment:
rm and just rely on requestKeyedData throwing exception.
It's racy to check here so the request below would have to handle it anyway

Reply:
done